### DISCUSSION TITLE
Can PWAsForFirefox be installed system wide so users on that system just have it automatically?

### DISCUSSION THREAD

**Discussion by joakim-tjernlund** [filips123/PWAsForFirefox#363] [Q&A] [answered]
If yes, how to do that?

**Comment by filips123** [comment]
Do you mean system wide installing just PWAsForFirefox program, or also all PWAs?

**Reply by joakim-tjernlund** [reply]
Just prepared, like chrome, so one only needs to point to teams.microsoft.com to start using Teams

**Reply by filips123** [reply]
The native program can be installed for all users (and it is by default). For the extension, you could set up a Firefox policy ([first link](https://support.mozilla.org/kb/deploying-firefox-with-extensions), [another link](https://github.com/mozilla/policy-templates#extensionsettings)) to automatically install the extension.

However, each user will still have to open the extension popup, accept the license agreement and install Firefox runtime. If you want to also install runtime globally, you will have to [modify the source code a bit](https://github.com/filips123/PWAsForFirefox/blob/main/native/src/components/runtime.rs#L123-L126), although I don't recommend this, as there might be issues with permissions and potential security problems.

**Reply by joakim-tjernlund** [reply]
What is the cfg(feature = "portable") ? Can I set it so I don't need to modify src?
This is on Gentoo so pwasforfirefox is built form source.

```
            if #[cfg(feature = "portable")] {
                // When compiling in PortableApps.com mode, the runtime is installed to <root>/App/PWAsForFirefox/runtime
                // This is used to comply with PortableApps.com guidelines about not having binary files in <root>/Data
                let directory = dirs.sysdata.join("runtime");
            } else {
                // Runtime is installed to a user-specific location along with other project files
                // If you want to overwrite locations of all project files, check `directories.rs`
                // If you want to overwrite only runtime location, replace the below line
                let directory = dirs.userdata.join("runtime");
            }
        }
```

**Reply by filips123** [reply]
It's for building the program with PortableApps.com support on Windows. It's also used in [a few other places](https://github.com/search?q=repo%3Afilips123%2FPWAsForFirefox+%22feature+%3D+%5C%22portable%5C%22%22&type=code), but only in combination with building on Windows, so if you set it on Linux, this is the only place that it *currently* affects. However, I cannot guarantee that this will also be the case in the future (in case I also add some portable versions for Linux or similar).

You can enable it using `cargo build --release --features portable`.

**Reply by joakim-tjernlund** [reply]
Perhaps FFPWA could search user dir first and fallback to system dir?
The it would just work automagically

**Reply by filips123** [reply]
Maybe I can try this. However, it needs to be a bit more complicated than that, as `firefoxpwa runtime install` should still install the runtime to the user dir.

**Comment by filips123** [comment, marked as answer]
This will be available in 2.8.0. The user runtime will always be used if it exists. If it doesn't exist, but the system runtime exists, it will be used instead. However, if none exist yet, `firefoxpwa runtime install` will install runtime to the user. Also, you will need to make sure that the runtime directory is writeable by all users that use PWAs, otherwise, patching probably won't work.

**Reply by joakim-tjernlund** [reply]
Thanks, can the Gentoo ebuild support an USE flag to also install a system FF PWA .xpi?

**Reply by filips123** [reply]
I don't know if Firefox still supports loading extensions directory from local files. I think they disabled this some time ago, but I'm not sure.

I found [this page](https://extensionworkshop.com/documentation/publish/distribute-sideloading/). It has a warning that:

> Starting in Firefox 73, it will no longer be possible to have an extension be automatically installed as part of another application install. See the [Add-ons Blog](https://blog.mozilla.org/addons/2019/10/31/firefox-to-discontinue-sideloaded-extensions/) for more information.

However, it also says below that:

> This method of add-on installation involves copying the add-on into one of the standard extension folders on the user's computer. Once copied, the next time Firefox launches the add-on will be installed. By default, the user will be asked to approve the installation, and if the user approves, the add-on will be installed and automatically loaded for subsequent launches.

So, maybe this option is still possible, but requires manual approval by each user.

Can you follow those steps to manually copy XPI file to the required location and test if Firefox can load it? If it works, I can try to add this to ebuild.

**Reply by joakim-tjernlund** [reply]
The LANG pack gentoo firefox ebuild install at:
```
/usr/lib64/firefox/distribution/extensions/langpack-sv-SE@firefox.mozilla.org.xpi
/usr/lib64/firefox/distribution/extensions/langpack-en-GB@firefox.mozilla.org.xpi
```
just work without any approval.

I will give the manual copy a try soon
